internal/command/init.go

```diff
@@ -874,6 +875,32 @@ func (c *InitCommand) getProviders(config *configs.Config, state *states.State,
 			return true, false, diags
 		}
 
+		// Jump in here and add a warning if any of the providers we've validated
+		// have only a single checksum. This usually means we didn't manage to
+		// pull checksums for all architectures remotely and only generated the
+		// checksum locally for the current architecture. There is a simple fix
+		// for this so terraform can print out some help.
+
+		var incompleteProviders []string
+		for provider, locks := range newLocks.AllProviders() {
+			if len(locks.AllHashes()) == 1 {
+				incompleteProviders = append(incompleteProviders, provider.ForDisplay())
+			}
+		}
+
+		if len(incompleteProviders) > 0 {
+			// We don't really care about the order here, we just want the output to be deterministic.
+			sort.Slice(incompleteProviders, func(i, j int) bool {
+				return incompleteProviders[i] < incompleteProviders[j]
+			})
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Warning,
+				"Incomplete validation for providers",
+				fmt.Sprintf(
+					"Terraform has only been able to record incomplete checksums for some providers (%s) so you may encounter problems when the lock file is used on machines with a different architecture. This can normally be fixed by running `terraform providers lock -platform=os_arch` for all target architectures.",
```

Comment:
I have some copyedits of this warning message which are admittedly subjective but my idea here is firstly to make this "sound like" the Terraform UI voice (which isn't defined strongly, but I think you'll get a sense for it over time) and also to give a little more information about what Terraform just did:

```
Warning: Incomplete lock file information for providers

Due to your customized provider installation methods, Terraform was forced
to calculate lock file checksums locally for the following providers:
  - hashicorp/aws
  - hashicorp/null
  - somethingelse/somethingelse

The current .terraform.lock.hcl file only includes checksums for
darwin_arm64, so Terraform running on another platform will fail to
install these providers.

To calculate additional checksums for another platform, run:
  terraform providers lock -platform=linux_amd64
(where linux_amd64 is the platform to generate)
```

My thought process as I wrote this was:
- Subjectively, feels nicer to enumerate the providers as a "bullet list", rather than as a comma-separated list, because the list could potentially be quite long and contain long strings like `foo.example.com/bleep-bloop/bar-baz` that probably won't word-wrap nicely.
- State that this is caused by the customized installation methods, which hopefully hints at a potential way to avoid this problem altogether (use the default settings) but without implying that it's _wrong_ to use customized settings, since there are of course various good reasons to do so.
- Mention the lock file name specifically so it's easier to see that we're talking about the same file that Terraform presumably also mentioned elsewhere in this output.
- Mention the platform the user is currently using, honestly mainly just to give another example of what a platform string might look like, since this might be the user's first exposure to them.
- Show the copy-pasteable command line on a line of its own, to avoid readers being confused about whether the delimiters are important. (This is our typical style with other error and warning messages which include commands to run).

    I also kinda arbitrarily chose to use `linux_amd64` as a "real example" of a platform string, rather than the "os_arch" placeholder, in the hope of making it easier to see what to do by showing another actual platform string. I chose `linux_amd64` in recognition that a typical way this problem manifests is someone writing a configuration originally on a Windows or macOS workstation but then trying to run it "for real" in automation running in Linux containers.
- Note that our diagnostic renderer in the CLI will automatically wrap any line of text that has no leading whitespace, assuming it's a paragraph, but will _not_ wrap lines which start with leading whitespace. In the above example, that applies to both the "bullet points" for the provider addresses and to the copy-pasteable command line at the end.